[RISCV] Added the MIPS prefetch extensions for MIPS RV64 P8700.

ukalappa-mips opened the pull request:
the extension enabled with xmipscbop.

Please refer "MIPS RV64 P8700/P8700-F Multiprocessing System Programmer’s Guide" for more info on the extension at https://mips.com/wp-content/uploads/2025/06/P8700_Programmers_Reference_Manual_Rev1.84_5-31-2025.pdf

ukalappa-mips committed:
Added prefetch extensions for MIPS RV64 P8700 and enable with xmipscbop option.

Please refer "MIPS RV64 P8700/P8700-F Multiprocessing System Programmer’s Guide" for more info on the extension
at https://mips.com/wp-content/uploads/2025/06/P8700_Programmers_Reference_Manual_Rev1.84_5-31-2025.pdf

ukalappa-mips committed:
Updated mcpu for mips-p8700 and related testcase for xmipscbop extension.

topperc commented on llvm/lib/Target/RISCV/Disassembler/RISCVDisassembler.cpp:
```diff
@@ -535,6 +535,19 @@ static DecodeStatus decodeRTZArg(MCInst &Inst, uint32_t Imm, int64_t Address,
+
+  // Imm is a signed immediate, so sign extend it.
+  if (Imm & (1 << (Bits - 1)))
+    Imm |= ~((1LL << Bits) - 1);
```
Please use the SignExtend utility in MathExtras.h

topperc commented on llvm/lib/Target/RISCV/Disassembler/RISCVDisassembler.cpp:
```diff
@@ -535,6 +535,19 @@ static DecodeStatus decodeRTZArg(MCInst &Inst, uint32_t Imm, int64_t Address,
   return MCDisassembler::Success;
 }
+template <int Bits>
+static DecodeStatus DecodeSImm(MCInst &Inst, uint64_t Imm, uint64_t Address,
```
Is this function used? The immediate is unsigned 9 bit rights?

ukalappa-mips replied:
yes and decodeSImmOperand() is dead definition

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXMips.td:
```diff
@@ -103,9 +109,41 @@ class SWPFormat<dag outs, dag ins, string opcodestr, string argstr>
+
+let Predicates = [HasVendorXMIPSCBOP] in {
+  // Prefetch Data Write.
+  def : Pat<(prefetch(AddrRegImm9(XLenVT GPR:$rs1),uimm9:$imm9),
```
```suggestion
  def : Pat<(prefetch (AddrRegImm9 (XLenVT GPR:$rs1), uimm9:$imm9),
```

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXMips.td:
```diff
@@ -103,9 +109,41 @@ class SWPFormat<dag outs, dag ins, string opcodestr, string argstr>
+             (i32 1), timm, (i32 1)),
+            (MIPSPREFETCH GPR:$rs1, uimm9:$imm9, 9)>;
+  // Prefetch Data Read.
+  def : Pat<(prefetch(AddrRegImm9(XLenVT GPR:$rs1),uimm9:$imm9),
```
```suggestion
  def : Pat<(prefetch (AddrRegImm9 (XLenVT GPR:$rs1), uimm9:$imm9),
```

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXMips.td:
```diff
@@ -103,9 +109,41 @@ class SWPFormat<dag outs, dag ins, string opcodestr, string argstr>
 // MIPS extensions
 //===----------------------------------------------------------------------===//
+let Predicates = [HasVendorXMIPSCBOP] ,DecoderNamespace = "Xmipscbop" in {
+  def MIPSPREFETCH : Mips_prefetch_ri<(outs),(ins GPR:$rs1, uimm9:$imm9, uimm5:$hint),
```
```suggestion
  def MIPSPREFETCH : Mips_prefetch_ri<(outs), (ins GPR:$rs1, uimm9:$imm9, uimm5:$hint),
```

topperc replied:
Should be MIPS_PREFETCH. The underscore represents the `.` in the instruction mnemonic.

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXMips.td:
```diff
@@ -103,9 +109,41 @@ class SWPFormat<dag outs, dag ins, string opcodestr, string argstr>
 }
 
+// Prefetch format.
+let hasSideEffects = 0, mayLoad = 1,mayStore = 1 in
```
```suggestion
let hasSideEffects = 0, mayLoad = 1, mayStore = 1 in
```

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXMips.td:
```diff
@@ -103,9 +109,41 @@ class SWPFormat<dag outs, dag ins, string opcodestr, string argstr>
+  bits<5> hint;
+
+  let Inst{31 - 29} = 0b000;
+  let Inst{28 - 20} = imm9{8 - 0};
```
You don't need `{8 - 0} if `imm9` is declare as 9 bits

topperc commented on llvm/test/CodeGen/RISCV/xmips-cbop.ll:
```diff
@@ -0,0 +1,40 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
```
This file was not generated by the script. Either use the script or delete the comment. Preferably use the script.

ukalappa-mips replied:
sure @topperc  ..thank you

mshockwave commented on llvm/test/CodeGen/RISCV/xmips-cbop.ll:
```diff
@@ -0,0 +1,40 @@
+  ret void
+}
+
+declare void @llvm.prefetch.p0(ptr readonly captures(none), i32 immarg, i32 immarg, i32 immarg) 
```
declaring intrinsics is no longer mandatory, so you can remove this line.

ukalappa-mips replied:
sure.

mshockwave commented on llvm/test/CodeGen/RISCV/xmips-cbop.ll:
```diff
@@ -0,0 +1,40 @@
+; RUN: llc -mtriple=riscv64 -mattr=+xmipscbop -mattr=+m -verify-machineinstrs < %s \
+; RUN:   | FileCheck %s -check-prefix=RV64XMIPSPREFETCH
+
+define dso_local void @prefetch_read(ptr noundef %a) {
```
I don't think dso_local is needed

ukalappa-mips commented on llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:
```diff
@@ -2925,6 +2925,54 @@ bool RISCVDAGToDAGISel::SelectAddrRegImm(SDValue Addr, SDValue &Base,
+      int64_t CVal = cast<ConstantSDNode>(Addr.getOperand(1))->getSExtValue();
+      assert(!isUInt<9>(CVal) && "uimm9 not already handled?");
+
+      if (isUInt<12>(CVal)) {
```
@topperc  ,thank you again and logic was to handle the constants that fit with 11 bits ....

ukalappa-mips committed:
Merge branch 'llvm:main' into ukalappa_xmipscbop

ukalappa-mips commented on llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:
```diff
@@ -2925,6 +2925,54 @@ bool RISCVDAGToDAGISel::SelectAddrRegImm(SDValue Addr, SDValue &Base,
+    }
+  }
+  // Immediates more than 12 bits i.e LUI,ADDI,ADD
+  if (selectConstantAddr(CurDAG, DL, VT, Subtarget, Addr, Base, Offset,
```
yes and this doesn't make sense for us our implementation.

ukalappa-mips commented on llvm/lib/Target/RISCV/RISCVInstrInfoXMips.td:
```diff
@@ -103,9 +109,41 @@ class SWPFormat<dag outs, dag ins, string opcodestr, string argstr>
+  bits<5> rs1;
+  bits<5> hint;
+
+  let Inst{31 - 29} = 0b000;
```
sure ..will update

ukalappa-mips commented on llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:
```diff
@@ -2925,6 +2925,54 @@ bool RISCVDAGToDAGISel::SelectAddrRegImm(SDValue Addr, SDValue &Base,
+      return true;
+    }
+
+    // Handle with 12 bit ofset  immediates with ADDI.
```
Updated

ukalappa-mips commented on llvm/lib/Target/RISCV/Disassembler/RISCVDisassembler.cpp:
```diff
@@ -535,6 +535,19 @@ static DecodeStatus decodeRTZArg(MCInst &Inst, uint32_t Imm, int64_t Address,
+template <int Bits>
+static DecodeStatus DecodeSImm(MCInst &Inst, uint64_t Imm, uint64_t Address,
+                               const MCDisassembler *Decoder) {
+  if (Imm & ~((1LL << Bits) - 1))
```
N/A

ukalappa-mips commented on llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:
```diff
@@ -1523,6 +1524,10 @@ bool RISCVAsmParser::matchAndEmitInstruction(SMLoc IDLoc, unsigned &Opcode,
+  case Match_InvalidUImm9:
+    return generateImmOutOfRangeError(
+        Operands, ErrorInfo, 0, (1 << 9) - 1,
+        "immediate must be a multiple of 9 bytes in the range");
```
Update the diagnostics

ukalappa-mips committed:
Made the following changes like

  a)Refactored the code for proper indent with clang format.
  b)Emit ADDI for 11 bits unsigned immediate offset .
  c)Removed the DecodeSImm definition (not used).
  d)Updated the testcase.

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXMips.td:
```diff
@@ -103,9 +109,41 @@ class SWPFormat<dag outs, dag ins, string opcodestr, string argstr>
+let Predicates = [HasVendorXMIPSCBOP] in {
+  // Prefetch Data Write.
+  def : Pat<(prefetch (AddrRegImm9  (XLenVT GPR:$rs1), uimm9:$imm9),
+             (i32 1), timm, (i32 1)),
```
What happens if the last argument to prefetch is 0 instead of 1? Does it cause a crash?

ukalappa-mips replied:
yes and it error out like "Cannot Select" and default data cache (last arg 1)  fetch which expected for us  ,and __builtin_prefetch intrinsic defaults to data cache .... and currently the intrinsic doesn't allow to override .

ukalappa-mips replied:
Sure @topperc  and updated the changes ..

topperc commented:
> the extension enabled with xmipscop.

This part of your description has the extension name misspelled

topperc commented on llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:
```diff
@@ -2925,6 +2925,49 @@ bool RISCVDAGToDAGISel::SelectAddrRegImm(SDValue Addr, SDValue &Base,
+
+    // Handle with 12 bit offset with sign bit off with ADDI.
+    // For Immediate Range [0, 2047]
+    else if (Addr.getOpcode() == ISD::ADD &&
```
I do not believe this code to handle ADD is necessary. If you don't match the ADD, the normal isel process will create an ADDI. You don't need to do it.

ukalappa-mips replied:
yes you are right and thank you @topperc

ukalappa-mips commented on llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:
```diff
@@ -1523,6 +1524,10 @@ bool RISCVAsmParser::matchAndEmitInstruction(SMLoc IDLoc, unsigned &Opcode,
+  case Match_InvalidUImm9:
+    return generateImmOutOfRangeError(
+        Operands, ErrorInfo, 0, (1 << 9) - 1,
+        "immediate offset must be in the range[0-511]");
```
Sure

ukalappa-mips commented on llvm/test/CodeGen/RISCV/xmips-cbop.ll:
```diff
@@ -0,0 +1,32 @@
+; RUN: llc -mtriple=riscv32 -mattr=+xmipscbop -mattr=+m -verify-machineinstrs < %s \
```
Sure ..Will use UTC

ukalappa-mips committed:
The following changes made to addressed the comments  like

  a)Refactored the code for proper indent with clang format.
  b)Updated SelectAddrRegImm9 to fall back to already isel exist pattern for ADD.
  c)Used UTC (utils/update_llc_test_checks.py) to generate the FileCheck CHECKS
  d)Updated the llvm-mc testcase accordingly.

ukalappa-mips committed:
Merge branch 'main' into ukalappa_xmipscbop

ukalappa-mips commented:
thank you @topperc , @mshockwave  and @lenary  for the comments ,please look out for the updated changes ...

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -2742,6 +2742,9 @@ bool RISCVInstrInfo::verifyInstruction(const MachineInstr &MI,
         case RISCVOp::OPERAND_UIMM9_LSB000:
           Ok = isShiftedUInt<6, 3>(Imm);
           break;
+        case RISCVOp::OPERAND_UIMM9:
```
This is still unresolved

ukalappa-mips committed:
Made the following changes
	a)make NOP for Prefetch INST.
	b)updated the testcase.

ukalappa-mips committed:
Merge branch 'main' into ukalappa_xmipscbop

topperc reviewed (approved):
LGTM but please wait for @lenary to approve too

lenary reviewed (approved):
LGTM
